autoload/magit/utils.vim

```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
 	return ( index(s:submodule_list, a:dirname) != -1 )
 endfunction
 
-" s:magit_cd_cmd: plugin variable to choose lcd/cd command, 'lcd' if exists,
-" 'cd' otherwise
-let s:magit_cd_cmd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
-" magit#utils#lcd: helper function to lcd. use cd if lcd doesn't exists
-function! magit#utils#lcd(dir)
-	execute s:magit_cd_cmd . a:dir
+" magit#utils#chdir will change the directory respecting
+" local/tab-local/global directory settings.
+function! magit#utils#chdir(dir)
+  " This is a dirty hack to fix tcd breakages on neovim.
+  " Future work should be based on nvim API and has a `if has('nvim')` check.
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+  else
+    let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
+  endif
+  execute chdir . ' ' . a:dir
+endfunction
+
+" magit#utils#pushd_root goes to git root directory, saving current directory
+" prior to changing the cwd.
+function! magit#utils#pushd_root()
+  if has('nvim')
+    let cwd = haslocaldir(-1, 0) ? getcwd(-1, 0) : getcwd()
+  else
+    let cwd = getcwd()
+  endif
+
+  let root = magit#git#top_dir()
+
+  if root !=? cwd
+    call magit#utils#chdir(root)
+    let g:_vimagit_cwd = cwd
```

Comment:
`_vimait_cwd` variable must be local to the current buffer (user may have several magit session opened in one vim session).